### PR TITLE
auth_code: default to header based authorization

### PR DESCRIPTION
**Pull request by maletor**
http://tools.ietf.org/html/rfc6749#section-2.3.1

> Including the client credentials in the request-body using the two
>    parameters is NOT RECOMMENDED and SHOULD be limited to clients unable
>    to directly utilize the HTTP Basic authentication scheme (or other
>    password-based HTTP authentication schemes).  The parameters can only
>    be transmitted in the request-body and MUST NOT be included in the
>    request URI.

### CODE REVIEW

**Comment by aviscasillas** [comment]
Hi!
Thinking on a solution to send client credentials using basic authentication, I've found this PR. I see that these changes are basically what I'm looking for, and I'm wondering if it will go forward (I see the last comment is old, but the PR is still open :confused:).
Thx!

**Comment by recmend** [comment]
Any update on this? This impacts Yahoo OAuth 2.0 integration. Yahoo refuses to permit client credentials in the request body.

**Comment by stevenringo** [comment]
Please can this be merged?

**Comment by sferik** [comment]
> I leave it up to you how you want to handle it, but basically our options are to:
> 1. Include a deprecation message in the get_token method here and support it until some time in the future.
> 2. Remove support (as it currently is) and major or minor version bump depending on how you feel about supporting an undocumented feature.

My preference would be for option 1 to give users a easier migration path.

@maletor Would you be willing to update your patch to deprecate but continue to support the existing behavior?

That is my only objection to merging and releasing this patch.

**Comment by maletor** [comment]
Sure, I can update the PR to deprecate it. I'll look at it sometime today.

**Comment by sferik** [comment]
@maletor Awesome! Thanks!

**Comment by stevenringo** [comment]
These changes also need to filter through to `AccessToken#refresh!`, which still uses the `:request_body` method. This looks like it could be a messy (re)factoring as `Strategy::Base#client_params` has the logic that's required in `AccessToken`.

**Comment by stevenringo** [comment]
I would like to make changes to this PR, to add more address the issue I raised in the previous comment. Do I create a new PR that incorporates this PR's changes, plus mine, or is there a way I can update this PR?

**Comment by maletor** [comment]
Feel free to fork my branch of cherry pick my commits. I agree, if we are
going to deprecate, it's going to require significantly more changes.

**Comment by etisab** [comment]
Hello, I was wondering what the status of these changes are and if they will they be merged in the near future. We are also doing Yahoo! OAuth 2.0 integration which requires authorization credentials in the request header. Thanks

**Comment by maletor** [comment]
Any updates on this?

**Comment by josephpage** [comment]
[Xee API](https://github.com/xee-lab/documentation) also requires authorization credentials in the request header for both get and renew access token.
Need this to be merged ! 🙏  @sferik

**Comment by bjeanes** [comment]
I also have an OAuth provider which rejects authentication other than by `Basic` for exchanging tokens. Does this need help getting it over the line or is there a bigger picture change in mind instead?

**Comment by ozgg** [comment]
So... is there any chance this will be merged at all?

**Comment by sferik** [comment]
This pull request no longer merges cleanly.

File: lib/oauth2/strategy/base.rb [hunk 1]
```diff
@@ -11,6 +11,14 @@ def initialize(client)
       def client_params
         {'client_id' => @client.id, 'client_secret' => @client.secret}
       end
+
+      # Returns the Authorization header value for Basic Authentication
+      #
+      # @param [String] The client ID
+      # @param [String] the client secret
+      def authorization(client_id, client_secret)
+        'Basic ' + Base64.encode64(client_id + ':' + client_secret).gsub("\n", '')
```
Comment (bjeanes) [inline lib/oauth2/strategy/base.rb]: @divoxx I don't think that is correct in the case of the `Authorization` header, as it is Base64 encoded. I wasn't sure so I tested to see what `curl` does here. Furthermore, the RFC does not specify any encoding other than Base64: https://tools.ietf.org/html/rfc2617#section-2.

Comment (bjeanes) [inline lib/oauth2/strategy/base.rb]: In my experiments with authenticating with a few OAuth providers, it seems that at least some reject form-encoded credentials. I have one with a `!` in the password, which would get encoded as `%21`, but this is rejected.

**Comment by ozgg** [comment]
Should @maletor or anyone else with write access do this? Because there are providers that fail to work with current version of the gem without these changes.

**Comment by bjeanes** [comment]
@sferik @ozgg I am currently preparing a new PR. ETA ~1 day

**Comment by bjeanes** [comment]
(I also have another PR in prep to bring the lib into spec around [redirect_uri for access tokens](https://tools.ietf.org/html/rfc6749#section-4.1))

**Pull request closed without merging**